### DISCUSSION TITLE
Template integration

### DISCUSSION THREAD

**Discussion by derbadewaschl** [mampfes/ha_epex_spot#50] [Q&A] [answered]
Hi,

The new 2.0.0 version says you can use "the [Template Integration](https://www.home-assistant.io/integrations/template/) to create a sensor that shows the start time:".
I can't get it running. I copy/pasted the code from the README into a new helper in Settings/Devices/Helpers, only updating the trigger time to a time 2 minutes from now. Then I waited 3 minutes. But when I click the helper it says "Unavailable". 

Also, it is not clear to me how to use it in an automation. Can I trigger on sensor "Start Appliance"? How?

Any help would be very much appreciated.

I use ha 2023.9.2

Regards

derbadewaschl

**Comment by mr-p666** [comment, marked as answer]
Hej,
I'm still new to HASS, but I don't think it's able to implement a template sensor as a helper like you'd like.
I implemented it directly into my configuration.yaml file and used a second trigger when starting HASS:
```
template:
  - trigger:
      - platform: time
        at: "00:00:00"
      - platform: homeassistant
        event: start
    action:
      - service: epex_spot.get_lowest_price_interval
        data:
          earliest_start: "00:00:00"
          latest_end: "23:59:59"
          duration:
            hours: 3
        response_variable: resp
    sensor:
      - name: "EPEX Start 24h Low Period"
        device_class: timestamp
        state: "{{ resp.start is defined and resp.start }}"
```
Afterwards you can use the sensor.epex_start_24h_low_period as a time triggered sensor in your automations.

**Reply by derbadewaschl** [reply]
I tested this for 3 days now and it worked on day 1: The water heater was turned on 12-14h on Oct 11. 

On Oct 12, however, the cheapest two hour period was 00:00 - 2:00 and my water heater was NOT turned on at 00:00. I suppose this is because 00:00 is also the trigger for calling get_lowest_price_interval. So I changed the trigger time to 23:30. 

Today, the cheapest time was 12:00 - 14:00, but again the water heater was not turned on. Looking at the history of "EPEX Start 24h Low Period", I can see that the value was stuck at Oct 12 at 00:00. I think get_lowest_price_interval returns the cheapest time period in all of the available data, not just the future hours. I guess at 23:30 there was still "Oct 12 00:00" in sensor.epex_spot_data_net_price and was again found as the cheapest start point.

**Reply by mampfes** [reply]
@derbadewaschl What did you set a earliest_start_time and latest_end_time, and when did you call the service? This all affects the returned values.
Your change from 00:00 to 23:30 shouldn't make a difference to the returned values, and therefore to the template sensor value. But it could make a difference in your automation, this depends how you trigger it.

**Reply by derbadewaschl** [reply]
The only thing I changed from the code you posted above is the duration (2 instead of 3).
I think this is what happened:
On 12 Oct 23:30 we had following values in sensor.epex_spot_data_net_price:
12 Oct 00:00 - 01:00: 9ct
12 Oct 01:00 - 02:00: 10ct
... (all prices >12ct)
13 Oct 12:00 - 13:00: 11ct
13 Oct 13:00 - 14:00: 12ct
... (all prices >12ct)
13 Oct 21:00 - 22:00: 20ct

When get_lowest_price_interval was called on 12 Oct 23:30, it returned 12 Oct 00:00 - 02:00, because it was the cheapest option in the available data. I guess the function does not check whether a value is in the past, does it?

**Reply by mampfes** [reply]
> I guess the function does not check whether a value is in the past, does it?

Correct. It assumes that earliest_start_time belongs to the current day, even if it is in the past.

And now I see your point. What you could do as a workaround is to set the earliest_start_time to 00:05, and the Trigger time to 00:00.

I will think of a solution for your problem.

**Reply by mr-p666** [reply]
It would be great, if it's possible to define a relative time (or timestamp) as earliest_start and latest_end.
eg. "+09:00:00"
Otherwise a config switch should do the trick:
eg. only_upcoming: true
With the first option you can also make >24h time windows (beginning at 10 or 11pm today end 12pm tommorow) because the lowest period can already starts at 10 or 11pm.

For now I use the following to achieve the same:
```
- trigger:
    - platform: time
      at: "21:55:00"
  action:
    - service: epex_spot.get_lowest_price_interval
      data:
        earliest_start: "22:00:00"
        duration:
          hours: 3
      response_variable: resp
  sensor:
    - name: "EPEX Start 26h Low Period"
      device_class: timestamp
      state: "{{ resp.start is defined and resp.start }}"
```

**Reply by mampfes** [reply]
Please check the main branch. You can now set an offset to get the lowest/highest price interval for e.g. tomorrow.

Please let me know if this fixes your problem. I would like to wait before I release it until I receive your feedback.

**Reply by derbadewaschl** [reply]
Hi,

Thank you so much. 
Here's what I configured now:
```
template:
  - trigger:
      - platform: time
        at: "21:30:00"
      - platform: homeassistant
        event: start
    action:
      - service: epex_spot.get_lowest_price_interval
        data:
          duration: "02:00"
          earliest_start: "22:00:00"
          latest_end: "15:00:00"
          latest_end_post: 1
        response_variable: resp
    sensor:
      - name: "EPEX Start 24h Low Period"
        device_class: timestamp
        state: "{{ resp.start is defined and resp.start }}"
```

Rationale: I am using this to keep the house at 10°C while I am away. Sometimes (not often, though) the cheapest period is at 22:00, I don't want to miss that. But I wouldn't want the heating to turn on at 22:00 and then again at 2:00 and then not at all until 22:00 the next day. 

I'll test this for the next couple of days and will let you know  how it went.

If you are open to suggestions to make the configuration easier, here is a proposal, how to configure the same thing as above:
```
          duration: "02:00"
          within: "17:00"
```

**Reply by mampfes** [reply]
@derbadewaschl If I understand your proposal correctly, then there is one major difference to the current solution: In your proposal, you have to call the service at the right time. Otherwise you will get a different result. That also means that you can't call the service upfront - which might be acceptable, but this depends on your setup.
Therefore I implemented the solution with the absolute timestamps. They are (to a certain extend) independent of the time the service is called.

**Reply by mr-p666** [reply]
I updated to the main branch (HASS is also up to date) and tried the following service call:
```
- trigger:
    - platform: time
      at: "20:30:00"
    - platform: homeassistant
      event: start
  action:
    - service: epex_spot.get_lowest_price_interval
      data:
        earliest_start: "06:00:00"
        earliest_start_post: 1
        latest_end: "22:00:00"
        latest_end_post: 1
        duration:
          hours: 3
      response_variable: resp
  sensor:
    - name: "EPEX Start Daytime Low Period"
      device_class: timestamp
      state: "{{ resp.start is defined and resp.start }}"
```
But it doesn't work. I found the following In the logfile:
```
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/template/coordinator.py", line 89, in _handle_triggered
    script_result = await self._script.async_run(run_variables, context)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 1578, in async_run
    return await asyncio.shield(run.async_run())
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 420, in async_run
    await self._async_step(log_exceptions=False)
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 470, in _async_step
    self._handle_exception(
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 493, in _handle_exception
    raise exception
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 468, in _async_step
    await getattr(self, handler)()
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 704, in _async_call_service_step
    response_data = await self._async_run_long_action(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 666, in _async_run_long_action
    return long_task.result()
           ^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/core.py", line 2012, in async_call
    response_data = await coro
                    ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/core.py", line 2049, in _execute_service
    return await target(service_call)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/epex_spot/__init__.py", line 113, in get_lowest_price_interval
    return _find_extreme_price_interval(call, lambda a, b: a < b)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/epex_spot/__init__.py", line 149, in _find_extreme_price_interval
    return coordinator.source.find_extreme_price_interval(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/epex_spot/SourceShell.py", line 144, in find_extreme_price_interval
    start_times = get_start_times(
                  ^^^^^^^^^^^^^^^^
  File "/config/custom_components/epex_spot/extreme_price_interval.py", line 145, in get_start_times
    raise ValueError("latest_end is earlier or equal than earliest_start")
ValueError: latest_end is earlier or equal than earliest_start
```
Would be great if someone could help me with this. :smile:

**Reply by mampfes** [reply]
Ah, this happens if you call the service before data for the next day is available.

Fixed in 2.1.0

**Comment by derbadewaschl** [comment]
> @derbadewaschl If I understand your proposal correctly, then there is one major difference to the current solution: In your proposal, you have to call the service at the right time. Otherwise you will get a different result. That also means that you can't call the service upfront - which might be acceptable, but this depends on your setup. Therefore I implemented the solution with the absolute timestamps. They are (to a certain extend) independent of the time the service is called.

I was not aware that there is a use case, which is independent of the time when you call it, as I am using it with the code that you provided in this thread. What's the alternative?

So far the new implementation (commit a6603eb) is working fine. Thanks again.
